runtime/j9vm/asgct.cpp

```diff
@@ -0,0 +1,158 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2021 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "j9.h"
+#include "j9protos.h"
+#include "ut_j9scar.h"
+#include "rommeth.h"
+
+#if defined(LINUX) && defined(J9VM_ARCH_X86) && defined(J9VM_ENV_DATA64)
+#include <ucontext.h>
+#define ASGCT_SUPPORTED
+#endif /* defined(LINUX) && defined(J9VM_ARCH_X86) && defined(J9VM_ENV_DATA64) */
+
+extern "C" {
+
+enum {
+ticks_no_Java_frame         =  0, // new thread
+ticks_no_class_load         = -1, // jmethodIds are not available
+ticks_GC_active             = -2, // GC action
+ticks_unknown_not_Java      = -3, // ¯\_(ツ)_/¯
+ticks_not_walkable_not_Java = -4, // ¯\_(ツ)_/¯
+ticks_unknown_Java          = -5, // ¯\_(ツ)_/¯
+ticks_not_walkable_Java     = -6, // ¯\_(ツ)_/¯
+ticks_unknown_state         = -7, // ¯\_(ツ)_/¯
+ticks_thread_exit           = -8, // dying thread
+ticks_deopt                 = -9, // mid-deopting code
+ticks_safepoint             = -10 // ¯\_(ツ)_/¯
+}; 
+
+typedef struct {
+	jint lineno;
+	jmethodID method_id;
+} ASGCT_CallFrame;
+
+typedef struct {
+	JNIEnv *env_id;
+	jint num_frames;
+	ASGCT_CallFrame *frames;
+} ASGCT_CallTrace;
+
+#if defined(ASGCT_SUPPORTED)
+
+extern J9JavaVM *BFUjavaVM;
+
+static UDATA
+asyncFrameIterator(J9VMThread * currentThread, J9StackWalkState * walkState)
+{
+	ASGCT_CallFrame *frame = (ASGCT_CallFrame*)walkState->userData1;
+	J9Method *method = walkState->method;
+	J9JNIMethodID *methodID = currentThread->javaVM->internalVMFunctions->getJNIMethodID(currentThread, method);
+	Assert_SC_notNull(methodID);
+	frame->method_id = (jmethodID)methodID;
+	J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+	if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccNative)) {
+		frame->lineno = -3;
```

Comment:
for these ones we can define something like `AGCT_LINE_NUMBER_NATIVE_METHOD` and `AGCT_LINE_NUMBER_NOT_AVAILABLE` for readability

Reply:
Done.